presto-main/src/main/java/com/facebook/presto/operator/scalar/TryFunction.java

```diff
@@ -0,0 +1,177 @@
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
+
+@Description("internal try function for desugaring TRY")
+@ScalarFunction(value = "$internal$try", hidden = true)
+public final class TryFunction
+{
+    private TryFunction() {}
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = long.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Long tryLong(@SqlType("function(T)") TryLongLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Long) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = double.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Double tryDouble(@SqlType("function(T)") TryDoubleLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Double) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = boolean.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Boolean tryBoolean(@SqlType("function(T)") TryBooleanLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Boolean) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = Slice.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Slice trySlice(@SqlType("function(T)") TrySliceLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Slice) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = Block.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Block tryBlock(@SqlType("function(T)") TryBlockLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Block) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = void.class)
+    @SqlNullable
+    @SqlType("T")
+    public static Void tryVoid(@SqlType("function(T)") TryVoidLambda function)
+    {
+        try {
+            return function.apply();
+        }
+        catch (PrestoException e) {
+            return (Void) tryFunctionExceptionHandler(e);
+        }
+    }
+
+    @FunctionalInterface
+    public interface TryLongLambda
+            extends LambdaFunctionInterface
+    {
+        Long apply();
```

Comment:
nit: I suggest to rename the interface name to `TryInvokeLongLambda` and the method name to `invoke` :)

Reply:
Well since our function name is `$internal$try`, I think current name is more consistent :) Let's see what others think about this.

As for method name, as per `LambdaBytecodeGenerator. getSingleApplyMethod())`, I don't think we could name it with other names.

Reply:
@shixuan-fan: You are right, I forgot the name must be `apply`.  my bad :D